### PR TITLE
feature: support Multi pk 

### PR DESCRIPTION
**Pull request by 81519434**
### Ⅰ. Describe what this PR did
pr content: To support multiple primary key.

It's a lot of work to support multiple primary keys for seata. Because this related to the base structure of seata.
I spend one month to do it. But It just works fine in mysql so far.
Because I don't have the time to compatible the seata with postgresql or oracle. And I‘m not good at PostgreSQL and oracle. I tried to compatible the seata with PostgreSQL and oracle, But I met some problems.
So I gave up.
It would be wonderful if someone can continue my work.

There are several bugs that I found when I implement this future. This fix is breaking changes.
1, Rollback will not work if update SQL changed the pk value. So I forbid the execution of this SQL.
2, We can't obtain the auto-increment value if the program run batch inserts SQL in the form of the statement. So I also forbid the execution of this SQL.

### CODE REVIEW

**Comment by 81519434** [comment]
I tested the future on my sample.
image
update\delete\insert\select for update   each of them has more than 8 test cases.

**Comment by l81893521** [comment]
Please fix the conflict.

**Comment by 81519434** [comment]
Ok, I will resolve the conflict this week.

**Comment by 81519434** [comment]
@l81893521  The conflict has been resolved.

**Comment by l81893521** [comment]
HI, I m so sorry about that we do not check this pr out for a long time. We decide check it now, so can you fix the conflict and keep in touch with me?

**Comment by l81893521** [comment]
If you like, you can find me in "DingTalk", my "DingTalk" code is "wpzfkml"

**Comment by 81519434** [comment]
已经加了好友了，请同意一下

**Comment by l81893521** [comment]
> 已经加了好友了，请同意一下 @l81893521

ok

File: rm/src/main/java/io/seata/rm/AbstractRMHandler.java [hunk 1]
```diff
@@ -105,6 +105,13 @@ protected void doBranchCommit(BranchCommitRequest request, BranchCommitResponse
             LOGGER.info("Branch commit result: " + status);
         }
 
+        for (GlobalTransactionHook hook : GlobalTransactionHookManager.popHooks(xid)) {
```
Comment (l81893521) [inline rm/src/main/java/io/seata/rm/AbstractRMHandler.java:108]: How about remove the GlobalTransactionHookManager, and make it to orther pr?

File: rm/src/main/java/io/seata/rm/DefaultResourceManager.java [hunk 2]
```diff
@@ -121,7 +121,7 @@ public void unregisterResource(Resource resource) {
 
     @Override
     public Map<String, Resource> getManagedResources() {
-        Map<String, Resource> allResource = new HashMap<>();
+        Map<String, Resource> allResource = new HashMap<>(10);
```
Comment (l81893521) [inline rm/src/main/java/io/seata/rm/DefaultResourceManager.java:124]: I think here keep it origin would be better.

Comment (81519434) [reply rm/src/main/java/io/seata/rm/DefaultResourceManager.java:124]: ok

**Comment by YangQiang-ChengDu** [comment]
希望早点支持复合主键

**Comment by jsbxyyx** [comment]
@81519434 attention check style.

**Comment by 81519434** [comment]
@jsbxyyx 
If the code logic is fine, I will optimize the code style in the end.

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java [hunk 3]
```diff
@@ -306,27 +316,106 @@ protected TableRecords queryCurrentRecords(Connection conn) throws SQLException
         return currentRecords;
     }
 
+    protected List<Field> getOrderedPkList(TableRecords image,Row row,String dbType) {
+        List<Field> pkFields = new ArrayList<>();
+        // To ensure the order of the pk, the order should based on getPrimaryKeyOnlyName.
+        List<String> pkColumnNameListByOrder = image.getTableMeta().getPrimaryKeyOnlyName();
```
Comment (l81893521) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java:322]: Why getPrimaryKeyOnlyName() here will be orderd?

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/KeywordCheckerFactory.java [hunk 4]
```diff
@@ -36,6 +37,10 @@
      * @return keyword checker
      */
     public static KeywordChecker getKeywordChecker(String dbType) {
+        if (JdbcConstants.H2.equalsIgnoreCase(dbType)) {
```
Comment (l81893521) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/KeywordCheckerFactory.java:40]: How about remove this? And create a new pr to support H2 keyword checker?

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java [hunk 5]
```diff
@@ -190,7 +193,7 @@ public static TableRecords buildRecords(TableMeta tmeta, ResultSet resultSet) th
                 int dataType = col.getDataType();
                 Field field = new Field();
                 field.setName(col.getColumnName());
-                if (tmeta.getPkName().equalsIgnoreCase(field.getName())) {
+                if (tmeta.getPrimaryKeyOnlyName().stream().anyMatch(e -> field.getName().equalsIgnoreCase(e))) {
```
Comment (l81893521) [inline rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:196]: How about change to
```
tmeta.getPrimaryKeyMap().containsKey(colName)
```

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java [hunk 6]
```diff
@@ -133,19 +135,20 @@ public void add(Row row) {
...
      * Pk rows list.
      *
-     * @return the list
+     * @return return a list. each element of list is a map,the map hold the pk column name as a key and field as the value
      */
-    public List<Field> pkRows() {
-        final String pkName = getTableMeta().getPkName();
-        List<Field> pkRows = new ArrayList<>();
+    public List<Map<String,Field>> pkRows() {
+        final List<String> pkNameList = getTableMeta().getPrimaryKeyOnlyName();
```
Comment (l81893521) [inline rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:141]: How about change to
```
final Map<String, ColumnMeta> primaryKeyMap = getTableMeta().getPrimaryKeyMap();
...
if (primaryKeyMap.containsKey(field.getName())) {

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java (hunk 3 above)
Comment (l81893521) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java:322]: I still have some question here.
1. you do not sort the list, why call it 'ordered'?
2.If you just want to get the same order of field in 'undo sql' and 'undo prepare', I think the order of field would be same when you get it in different time.

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java [hunk 7]
```diff
@@ -306,27 +316,106 @@ protected TableRecords queryCurrentRecords(Connection conn) throws SQLException
...
+        return whereStr.toString();
+
+    }
+    /**
+     * each pk is a condition.the result will like :" id in (?,?,?) and userCode in (?,?,?)"
+     * @param pkNameList
+     * @param pkRowValues  the kye of map is pk name ,and value of map is pk's value
+     * @return return where condition sql string.the sql can search all related records not just one.
+     */
+    protected String buildWhereConditionByPKs(List<String> pkNameList,Map<String,List<Field>> pkRowValues,Connection conn) throws SQLException {
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java:362]: (a1 and b1) or ( a1 and b2 ) or (a2 and b2)  <=>  A in (a1,a2) and B in (b1,b2) ?

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java [hunk 8]
```diff
@@ -306,27 +316,106 @@ protected TableRecords queryCurrentRecords(Connection conn) throws SQLException
...
-            List<Field> fields = undoRows.get(i).getFields();
+    protected Map<String,List<Field>> parsePkValues(List<Row> rows, List<String> pkNameList) {
+        List<Field> pkFieldList = new ArrayList<>();
+        for (int i = 0; i < rows.size(); i++) {
+            List<Field> fields = rows.get(i).getFields();
             if (fields != null) {
                 for (Field field : fields) {
-                    if (StringUtils.equalsIgnoreCase(pkName, field.getName())) {
-                        pkValues[i] = field.getValue();
+                    if (pkNameList.stream().anyMatch(e -> field.getName().equals(e))) {
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java:411]: ignoreCasesEqual?

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/KeywordCheckerFactory.java (hunk 4 above)
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/KeywordCheckerFactory.java:40]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java (hunk 5 above)
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:196]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java (hunk 6 above)
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/sql/struct/TableRecords.java:141]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java [hunk 9]
```diff
@@ -273,30 +280,33 @@ protected boolean dataValidationAndGoOn(Connection conn) throws SQLException {
...
         }
-        StringBuilder replace = new StringBuilder();
-        for (int i = 0; i < pkValues.length; i++) {
-            replace.append("?,");
-        }
         // build check sql
         String dbType = getDbType(conn);
-        String checkSQL = String.format(CHECK_SQL_TEMPLATE, sqlUndoLog.getTableName(),
-            tableMeta.getEscapePkName(dbType), replace.substring(0, replace.length() - 1));
+        String checkSQL = String.format(CHECK_SQL_TEMPLATE, ColumnUtils.addEscape(sqlUndoLog.getTableName(), dbType),
```
[the remark this answers is not in the document]
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java:293]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java (hunk 8 above)
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java:411]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/mysql/MySQLInsertExecutor.java [hunk 10]
```diff
@@ -63,19 +68,55 @@ public MySQLInsertExecutor(StatementProxy statementProxy, StatementCallback stat
...
         // PK is just auto generated
+        Map<String, List<Object>> pkValuesMap = new HashMap<>(8);
         Map<String, ColumnMeta> pkMetaMap = getTableMeta().getPrimaryKeyMap();
-        ColumnMeta pkMeta = pkMetaMap.values().iterator().next();
-        if (!pkMeta.isAutoincrement()) {
+        String autoColumnName = "";
+        for (String pkColumnName : pkMetaMap.keySet()) {
+            if (pkMetaMap.get(pkColumnName).isAutoincrement())
+            {
+                autoColumnName = pkColumnName;
```
[the remark this answers is not in the document]
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/mysql/MySQLInsertExecutor.java:115]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java [hunk 11]
```diff
@@ -306,27 +316,106 @@ protected TableRecords queryCurrentRecords(Connection conn) throws SQLException
...
+    protected List<Field> getOrderedPkList(TableRecords image,Row row,String dbType) {
+        List<Field> pkFields = new ArrayList<>();
+        // To ensure the order of the pk, the order should based on getPrimaryKeyOnlyName.
+        List<String> pkColumnNameListByOrder = image.getTableMeta().getPrimaryKeyOnlyName();
+        List<String> pkColumnNameListNoOrder = row.primaryKeys()
+                .stream()
+                .map(e -> ColumnUtils.delEscape(e.getName(),dbType))
+                .collect(Collectors.toList());
+        pkColumnNameListByOrder.forEach(pkName -> {
+            int pkIndex = pkColumnNameListNoOrder.indexOf(pkName);
```
[the remark this answers is not in the document]
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java:328]: The source of the column name is the same. 
You will find out that they are coming from Tablemate if you track the invocation chain.

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java (hunk 7 above)
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/AbstractUndoExecutor.java:362]: ok

File: rm/src/main/java/io/seata/rm/AbstractRMHandler.java (hunk 1 above)
Comment (81519434) [reply rm/src/main/java/io/seata/rm/AbstractRMHandler.java]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLUndoInsertExecutor.java [hunk 12]
```diff
@@ -54,18 +56,26 @@ protected String buildUndoSQL() {
...
+        }
+    }
+
+    private String generateDeleteSql(List<Row> rows,TableRecords afterImage) {
+        KeywordChecker keywordChecker = KeywordCheckerFactory.getKeywordChecker(JdbcConstants.MYSQL);
+        List<String> pkNameList = getOrderedPkList(afterImage,rows.get(0),JdbcConstants.MYSQL)
+                .stream().map(e -> e.getName())
+                .collect(Collectors.toList());
+        String whereSql = buildWhereConditionByPKs(pkNameList,keywordChecker);
+        return String.format(DELETE_SQL_TEMPLATE,keywordChecker.checkAndReplace(sqlUndoLog.getTableName()), whereSql);
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLUndoInsertExecutor.java:78]: sqlUndoLog.getTableName() always has escape.

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLUndoUpdateExecutor.java [hunk 13]
```diff
@@ -47,21 +49,27 @@
...
             .collect(Collectors.joining(", "));
-        return String.format(UPDATE_SQL_TEMPLATE, sqlUndoLog.getTableName(), updateColumns,
-                ColumnUtils.addEscape(pkField.getName(), JdbcConstants.MYSQL));
+
+        List<String> pkNameList = getOrderedPkList(beforeImage,row,JdbcConstants.MYSQL)
+                .stream().map(e -> e.getName())
+                .collect(Collectors.toList());
+        String whereSql = buildWhereConditionByPKs(pkNameList,keywordChecker);
+
+        return String.format(UPDATE_SQL_TEMPLATE, keywordChecker.checkAndReplace(sqlUndoLog.getTableName()),updateColumns, whereSql);
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLUndoUpdateExecutor.java:72]: sqlUndoLog.getTableName() always has escape.

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/oracle/OracleUndoInsertExecutor.java [hunk 14]
```diff
@@ -49,18 +51,27 @@ protected String buildUndoSQL() {
...
+    }
+
+
+    private String generateDeleteSql(List<Row> rows,TableRecords afterImage) {
+        KeywordChecker keywordChecker = KeywordCheckerFactory.getKeywordChecker(JdbcConstants.ORACLE);
+        List<String> pkNameList = getOrderedPkList(afterImage,rows.get(0),JdbcConstants.ORACLE)
+                .stream().map(e -> e.getName())
+                .collect(Collectors.toList());
+        String whereSql = buildWhereConditionByPKs(pkNameList,keywordChecker);
+        return String.format(DELETE_SQL_TEMPLATE,keywordChecker.checkAndReplace(sqlUndoLog.getTableName()), whereSql);
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/oracle/OracleUndoInsertExecutor.java:74]: sqlUndoLog.getTableName() always has escape.

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/oracle/OracleUndoUpdateExecutor.java [hunk 15]
```diff
@@ -36,28 +38,33 @@
...
-                ColumnUtils.addEscape(pkField.getName(), JdbcConstants.ORACLE));
+                .map(field -> ColumnUtils.addEscape(field.getName(), JdbcConstants.ORACLE) + " = ?")
+                .collect(Collectors.joining(", "));
+
+        List<String> pkNameList = getOrderedPkList(beforeImage,row,JdbcConstants.ORACLE)
+                .stream().map(e -> e.getName())
+                .collect(Collectors.toList());
+        String whereSql = buildWhereConditionByPKs(pkNameList,keywordChecker);
+
+        return String.format(UPDATE_SQL_TEMPLATE, keywordChecker.checkAndReplace(sqlUndoLog.getTableName()),updateColumns, whereSql);
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/oracle/OracleUndoUpdateExecutor.java:67]: sqlUndoLog.getTableName() always has escape.

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/postgresql/PostgresqlUndoInsertExecutor.java [hunk 16]
```diff
@@ -49,18 +51,26 @@ protected String buildUndoSQL() {
...
+    }
+
+
+    private String generateDeleteSql(List<Row> rows,TableRecords afterImage) {
+        KeywordChecker keywordChecker = KeywordCheckerFactory.getKeywordChecker(JdbcConstants.POSTGRESQL);
+        List<String> pkNameList = getOrderedPkList(afterImage,rows.get(0),JdbcConstants.POSTGRESQL)
+                .stream().map(e -> e.getName())
+                .collect(Collectors.toList());
+        String whereSql = buildWhereConditionByPKs(pkNameList,keywordChecker);
+        return String.format(DELETE_SQL_TEMPLATE,keywordChecker.checkAndReplace(sqlUndoLog.getTableName()), whereSql);
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/postgresql/PostgresqlUndoInsertExecutor.java:73]: sqlUndoLog.getTableName() always has escape.

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/postgresql/PostgresqlUndoUpdateExecutor.java [hunk 17]
```diff
@@ -34,27 +36,33 @@
...
                 .collect(Collectors.joining(", "));
-        return String.format(UPDATE_SQL_TEMPLATE, sqlUndoLog.getTableName(), updateColumns,
-                ColumnUtils.addEscape(pkField.getName(), JdbcConstants.POSTGRESQL));
+
+        List<String> pkNameList = getOrderedPkList(beforeImage,row,JdbcConstants.POSTGRESQL)
+                .stream().map(e -> e.getName())
+                .collect(Collectors.toList());
+        String whereSql = buildWhereConditionByPKs(pkNameList,keywordChecker);
+
+        return String.format(UPDATE_SQL_TEMPLATE, keywordChecker.checkAndReplace(sqlUndoLog.getTableName()),updateColumns, whereSql);
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/undo/postgresql/PostgresqlUndoUpdateExecutor.java:65]: sqlUndoLog.getTableName() always has escape.

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java [hunk 18]
```diff
@@ -120,18 +124,34 @@ public T execute(Object... args) throws Throwable {
...
      * @return the string
      * @throws SQLException the sql exception
      */
-    protected String buildWhereConditionByPKs(List<Field> pkRows) throws SQLException {
-        StringJoiner whereConditionAppender = new StringJoiner(",", getColumnNameInSQL(pkRows.get(0).getName()) + " in (", ")");
-        for (Field field : pkRows) {
-            whereConditionAppender.add("?");
+    protected String buildWhereConditionByPKs(List<Map<String,Field>> pkRows) throws SQLException {
+        StringBuilder sql = new StringBuilder();
+        List<String> pkColumnNameList = getTableMeta().getPrimaryKeyOnlyName();
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java:135]: same  as  another `buildWhereConditionByPKs `

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java [hunk 19]
```diff
@@ -331,22 +388,45 @@ protected TableRecords buildTableRecords(TableMeta tableMeta, String selectSQL,
...
+        StringBuilder sql =  new StringBuilder()
+                .append("SELECT * FROM ")
+                .append(getFromTableInSQL())
+                .append(" WHERE ");
+
+        List<String> pkColumnNameList = getTableMeta().getPrimaryKeyOnlyName();
+        for (int i = 0;i < pkColumnNameList.size();i++)
+        {
+            if (i > 0)
+            {
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java:406]: same as another buildWhereConditionByPKs

File: rm-datasource/src/main/java/io/seata/rm/datasource/DataCompareUtils.java [hunk 20]
```diff
@@ -180,20 +182,28 @@ private static void convertType(Field f0, Field f1) {
...
+                    .sorted(Comparator.comparing(Field::getName))
+                    .collect(Collectors.toList());
             // {uppercase fieldName : field}
             Map<String, Field> colsMap = new HashMap<>();
-            String rowKey = null;
-            for (int j = 0; j < row.getFields().size(); j++) {
-                Field field = row.getFields().get(j);
-                if (field.getName().equalsIgnoreCase(primaryKey)) {
-                    rowKey = String.valueOf(field.getValue());
+            String rowKey = new String();
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/DataCompareUtils.java:195]: StringUtils.empty

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLUndoInsertExecutor.java (hunk 12 above)
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLUndoInsertExecutor.java:78]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLUndoUpdateExecutor.java (hunk 13 above)
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/mysql/MySQLUndoUpdateExecutor.java:72]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/oracle/OracleUndoInsertExecutor.java (hunk 14 above)
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/oracle/OracleUndoInsertExecutor.java:74]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/oracle/OracleUndoUpdateExecutor.java (hunk 15 above)
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/oracle/OracleUndoUpdateExecutor.java:67]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/postgresql/PostgresqlUndoInsertExecutor.java (hunk 16 above)
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/postgresql/PostgresqlUndoInsertExecutor.java:73]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/undo/postgresql/PostgresqlUndoUpdateExecutor.java (hunk 17 above)
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/undo/postgresql/PostgresqlUndoUpdateExecutor.java:65]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java (hunk 19 above)
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java:406]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java (hunk 18 above)
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseTransactionalExecutor.java:135]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/DataCompareUtils.java (hunk 20 above)
Comment (81519434) [reply rm-datasource/src/main/java/io/seata/rm/datasource/DataCompareUtils.java:195]: ok

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/AbstractDMLBaseExecutor.java [hunk 21]
```diff
@@ -83,6 +85,10 @@ public T doExecute(Object... args) throws Throwable {
      * @throws Exception the exception
      */
     protected T executeAutoCommitFalse(Object[] args) throws Exception {
+        if (!JdbcConstants.MYSQL.equalsIgnoreCase(getDbType()) && getTableMeta().getPrimaryKeyOnlyName().size() > 1)
```
Comment (slievrly) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/AbstractDMLBaseExecutor.java:88]: why only  mysql?

File: rm-datasource/src/test/java/io/seata/rm/datasource/undo/h2/keyword/H2KeywordChecker.java [hunk 22]
```diff
@@ -0,0 +1,12 @@
+package io.seata.rm.datasource.undo.h2.keyword;
+
+import io.seata.common.loader.LoadLevel;
+import io.seata.rm.datasource.undo.mysql.keyword.MySQLKeywordChecker;
+import io.seata.sqlparser.util.JdbcConstants;
+
+/**
+ * @author JerryYin
+ */
+@LoadLevel(name = JdbcConstants.H2)
```
Comment (slievrly) [inline rm-datasource/src/test/java/io/seata/rm/datasource/undo/h2/keyword/H2KeywordChecker.java:10]: Is it in the wrong place?

File: rm-datasource/pom.xml [hunk 23]
```diff
@@ -83,7 +83,6 @@
         <dependency>
             <groupId>com.google.guava</groupId>
```
Comment (slievrly) [inline rm-datasource/pom.xml:84]: remove guava dependency， cause  rm already add.

**Comment by sabersword** [comment]
Why does the feature support multi pk only for mysql? 
Is there any incompatible grammer with oracle?
Thank you for the details !

**Comment by 81519434** [comment]
Sorry @sabersword , My company only uses MySQL. And I don't have time to support other databases.
You can create PR to support other databases if you need this feature.

**Comment by a364176773** [comment]
请签署Seata社区 cla  https://cla-assistant.io/seata/seata  谢谢配合

**Pull request merged**